src/Http/Http/src/Timeouts/RequestTimeoutsMiddleware.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Http.Timeouts;
+
+internal sealed class RequestTimeoutsMiddleware
+{
+    private readonly RequestDelegate _next;
+    private readonly IOptionsMonitor<RequestTimeoutOptions> _options;
+
+    public RequestTimeoutsMiddleware(RequestDelegate next, IOptionsMonitor<RequestTimeoutOptions> options)
+    {
+        _next = next;
+        _options = options;
+    }
+
+    public Task Invoke(HttpContext context)
+    {
+        var endpoint = context.GetEndpoint();
+        var timeoutMetadata = endpoint?.Metadata.GetRequiredMetadata<RequestTimeoutAttribute>();
+        var policyMetadata = endpoint?.Metadata.GetRequiredMetadata<RequestTimeoutPolicy>();
+
+        var options = _options.CurrentValue;
+
+        if (timeoutMetadata is null && policyMetadata?.Timeout is null && options.DefaultTimeout is null)
+        {
+            return _next(context);
+        }
+
+        var disableMetadata = endpoint?.Metadata.GetRequiredMetadata<DisableRequestTimeoutAttribute>();
+        if (disableMetadata is not null)
+        {
+            return _next(context);
+        }
+
+        TimeSpan? timespan;
+        if (policyMetadata is not null)
+        {
+            timespan = policyMetadata.Timeout;
+        }
+        else if (timeoutMetadata is not null)
+        {
+            if (timeoutMetadata.Timeout is not null)
+            {
+                timespan = timeoutMetadata.Timeout.Value;
+            }
+            else
+            {
+                if (options.Policies.TryGetValue(timeoutMetadata.PolicyName!, out var policy))
+                {
+                    timespan = policy.Timeout;
+                }
+                else
+                {
+                    // policy does not exists. what to do now!?
+                    throw new Exception();
+                }
+            }
+        }
+        else
+        {
+            timespan = options.DefaultTimeout;
+        }
+
+        if (timespan is null || timespan == Timeout.InfiniteTimeSpan)
+        {
+            return _next(context);
+        }
+
+        return SetTimeoutAsync();
+
+        async Task SetTimeoutAsync()
+        {
+            var originalToken = context.RequestAborted;
+            var timeoutCts = new CancellationTokenSource(timespan.Value);
```

Comment:
A `TimeSpan` can be very large. `CancellationTokenSource` timeout only supports up to 2^32 milliseconds. Either support up to `TimeSpan.MaxValue` (you can do this using a Timer that triggers multiple internals) or provide validation of the timeout.

Reply:
Isn't that ~49 days? That seems sufficient :-).